downstairs/src/extent_inner_raw.rs

```diff
@@ -0,0 +1,1325 @@
+            cdt::extent__write__get__hashes__start!(|| {
+                (job_id.0, self.extent_number, writes.len() as u64)
+            });
+            let mut write_run_start = 0;
+            while write_run_start < writes.len() {
+                let first_write = writes[write_run_start];
+
+                // Starting from the first write in the potential run, we scan
+                // forward until we find a write with a block that isn't
+                // contiguous with the request before it. Since we're counting
+                // pairs, and the number of pairs is one less than the number of
+                // writes, we need to add 1 to get our actual run length.
+                let n_contiguous_writes = writes[write_run_start..]
+                    .windows(2)
+                    .take_while(|wr_pair| {
+                        wr_pair[0].offset.value + 1 == wr_pair[1].offset.value
+                    })
+                    .count()
+                    + 1;
+
+                // Query hashes for the write range.
+                let block_contexts = self.get_block_contexts(
+                    first_write.offset.value,
+                    n_contiguous_writes as u64,
+                )?;
+
+                for (i, block_contexts) in block_contexts.iter().enumerate() {
+                    if block_contexts.is_some() {
+                        let _ = writes_to_skip
+                            .insert(i as u64 + first_write.offset.value);
+                    }
+                }
+
+                write_run_start += n_contiguous_writes;
+            }
+            cdt::extent__write__get__hashes__done!(|| {
+                (job_id.0, self.extent_number, writes.len() as u64)
+            });
+
+            if writes_to_skip.len() == writes.len() {
+                // Nothing to do
+                return Ok(());
+            }
+        }
+
+        self.set_dirty()?;
+
+        // Write all the metadata to the raw file, at the end
+        //
+        // TODO right now we're including the integrity_hash() time in the
+        // measured time.  Is it small enough to be ignored?
+        cdt::extent__write__raw__context__insert__start!(|| {
+            (job_id.0, self.extent_number, writes.len() as u64)
+        });
+
+        let mut pending_slots = vec![];
+        for write in writes {
+            if writes_to_skip.contains(&write.offset.value) {
+                pending_slots.push(None);
+                continue;
+            }
+
+            // TODO it would be nice if we could profile what % of time we're
+            // spending on hashes locally vs writing to disk
+            let on_disk_hash = integrity_hash(&[&write.data[..]]);
+
+            let next_slot =
+                self.set_block_context(&DownstairsBlockContext {
+                    block_context: write.block_context,
+                    block: write.offset.value,
+                    on_disk_hash,
+                })?;
```

Comment:
If any of the extent data writes below fail, then the downstairs will re-execute this whole function for the same `writes` argument. It looks like `set_block_context` is destructive for the current contexts, which means if this function is retried two times (failing each time during the extent data write) then it will overwrite the correct context for all N blocks in `writes`.

Reply:
This is a little subtle, but I think it's correct.

- Let's say that block 0 is using context slot A.
- `set_block_context` will write to context slot B, but _will not_ update `self.active_context`.
- If we then fail the extent data write, we will then actively read the data back from the file and check which context slot is valid
  - If the new data was written, according to reading back from the file (but then the write returned an error code?), we will update the context slot to slot B
  - If the new data wasn't written, we leave the context slot at slot A.  This means that we're in the same place that we started, where the data in the file corresponds to context slot A